### PR TITLE
Expert partitioner: userfrienly management of subvolumes (fate#320296)

### PR DESCRIPTION
**Pull request by ancorgs**
Tested manually.

### CODE REVIEW

File: src/lib/storage/shadowed_vol_helper.rb [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        list.concat(disk.fetch("partitions", []))
+      end
+    end
+
+    def root_part
+      partitions.detect {|p| p["mount"] == "/" }
+    end
+
+    def doable?
+      if root_part.nil?
```
Comment (lslezak) [inline src/lib/storage/shadowed_vol_helper.rb]: NP: use a guard clause here, it saves some indentation:
```ruby
return true unless root_part.nil?

log.info ...
```

File: src/lib/storage/shadowed_vol_helper.rb [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    # Copy of the root partition with all the shadowed subvolumes marked for
+    # deletion and with the previously deleted subvolumes restored if they are
+    # not shadowed anymore
+    #
+    # In order to make the recovery of previously deleted subvolumes possible,
+    # it stores the deleted volumes in a class variable. This list needs to be
+    # cleaned between subsequent calls to the expert partitioner.
+    # @see .cleanup
+    def root_partition
+      log.info " shadowed subvolumes"
```
Comment (lslezak) [inline src/lib/storage/shadowed_vol_helper.rb]: ??? I guess it's just a debug log which will be removed or improved...

File: src/lib/storage/shadowed_vol_helper.rb [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        elsif existing_subvol["delete"]
+          log.info "Rejecting subvol deletion"
+          existing_subvol["delete"] = false
+        end
+      end
+      clean_aborted_subvols
+    end
+
+    def aborted_subvols
+      @@aborted_subvols ||= []
```
Comment (lslezak) [inline src/lib/storage/shadowed_vol_helper.rb]: I use a singleton instance for keeping persistent data (https://github.com/yast/yast-registration/blob/master/src/lib/registration/storage.rb#L31), IMHO it's a bit more consistent with "normal" objects. But I'm not against class attributes, I simply do not know which one is better... :grey_question: 

But we should be consistent across YaST modules and use one pattern preferably... a topic for discussion?

File: test/fixtures/msdos-btrfs-shadowed-twice.ycp [hunk 4]
```diff
@@ -0,0 +1,246 @@
+$[
```
Comment (lslezak) [inline test/fixtures/msdos-btrfs-shadowed-twice.ycp]: NP: Could we get rid of the YCP data format? I know it's a bit out of scope of this PR but we should finally kill the last `*.ycp` zombies...

BTW there is a [`ycp2yml.rb` conversion tool](https://github.com/yast/yast-devtools/blob/master/ytools/yast2/ycp2yml.rb)...

File: src/lib/storage/shadowed_vol_helper.rb (hunk 1 above)
Comment (ancorgs) [inline src/lib/storage/shadowed_vol_helper.rb]: Actually I have the feeling we abuse guard clauses sometimes. For short cases like this, I personally find the `if..else` approach more readable. A matter of taste.

File: src/lib/storage/shadowed_vol_helper.rb (hunk 2 above)
Comment (ancorgs) [inline src/lib/storage/shadowed_vol_helper.rb]: Yep, it definitely looks like a leftover.

File: src/lib/storage/shadowed_vol_helper.rb (hunk 3 above)
Comment (ancorgs) [inline src/lib/storage/shadowed_vol_helper.rb]: I think you are right. The singleton pattern is more common in YaST codebase and, with the current shape of this class, there is no strong reason to not use singleton. I will change it

**Comment by lslezak** [comment]
Thanks, LGTM now! :+1:

**Comment by kobliha** [comment]
@ancorgs Great job! :+1: Thank you!

**Pull request merged**